mailbox/memory/src/main/java/org/apache/james/mailbox/inmemory/mail/InMemoryMailboxMapper.java

```diff
@@ -158,24 +158,24 @@ public void updateACL(Mailbox mailbox, MailboxACL.MailboxACLCommand mailboxACLCo
         mailbox.setACL(mailbox.getACL().apply(mailboxACLCommand));
     }
 
+    private synchronized Iterable<MailboxAnnotation> retrieveAllAnnotations(InMemoryId maiboxId) {
```

Comment:
You are mixing two types of synchronised....

Line 61 : 

```
private synchronized Iterable<MailboxAnnotation> retrieveAllAnnotations(InMemoryId maiboxId) {
```

Line 194 : 

```
synchronized(mailboxAnnotations) {
```

So update and retrieveAllAnnotations can execute concurrently.

You should choose one solution for you locking.

1/ Synchronyze on the methods
2/ Create something like this : 

```
   private final Object lock = new Object();

   public void any() {
       synchronize (lock) {
          ....
       }
   }
```